### DISCUSSION TITLE
Is this possible? Shader Based Normal Editing but in Malt?

### DISCUSSION THREAD

**Discussion by datOne23** [bnpr/Malt#281] [General]
[manipulating normals using shader](https://www.youtube.com/watch?v=B186kSYglEA)

There is this tutorial about manipulating normals but it's on Eevee and uses drivers for it to work. I've been wondering if this is possible in Malt but my brain just shutdowns doing technical things in Malt.

**Comment by panthavma** [comment]
I haven't watched the whole video, but this is possible.

The algorithm explained in the video seems to be doable directly from a vertex or fragment shader (preferably fragment for better precision), as you have access to the mesh normal through `IO_NORMAL` and the camera matrix through `CAMERA`. From there you could mix the mesh normals and the camera's back vector to get the outlined result.

**Comment by pragma37** [comment]
Sure, you can use drivers in Malt as well. The only gotcha is that you must use them from the Material properties, they don't work in nodes due to a Blender limitation.

That said, you can retrieve the camera direction from shaders.
There's a built-in view_direction function that returns the view to surface normal in world space.

If you want the camera direction at the center of the view you can retrieve it from an Inline Code node with this code:
```glsl
transform_normal(inverse(CAMERA), vec3(0,0,-1))
```

I haven't watched the full video, but I think the algorithm boils down to this:
```glsl
vec3 camera_facing_normals(sampler1D distribution)
{
    vec3 ortho_view_normal = transform_normal(inverse(CAMERA), vec3(0,0,1));
    float weight = dot(NORMAL, ortho_view_normal);
    weight = texture(distribution, max(weight, 0)).r;
    return normalize(mix(NORMAL, ortho_view_normal, weight));
}
```
I'm using a color ramp instead of curves because that's what Malt supports atm.

Remember to disable self shadows, otherwise you won't see too much difference.

**Reply by datOne23** [reply]
Im going to try this when I have enough free time. Turns out drivers work on the material properties.
